feat: upgrade to Apollo Client v4

hanihusam opened the pull request:
## Summary

This PR upgrades the storybook-addon-apollo-client to support Apollo Client v4, addressing compatibility issues with the latest version.

### Key Changes
- Updated `@apollo/client` from v3.13.8 to v4.0.5
- Fixed import paths for `MockedProvider` (now from `@apollo/client/testing/react`)
- Updated peer dependencies to support both v3 and v4 (`^3.0.0 || ^4.0.0`)
- Replaced `ApolloError` with `GraphQLError` in stories and documentation
- Updated import paths in components to use Apollo Client v4 structure
- Added proper TypeScript types for GraphQL queries

### Breaking Changes
- Import path for `MockedProvider` changed from `@apollo/client/testing` to `@apollo/client/testing/react`
- Some Apollo Client v3 specific imports may need updates

### Testing
- Updated story examples to use `GraphQLError` instead of `ApolloError`
- Verified components work with both hook-based and render prop patterns

Closes #135

hanihusam committed:
feat: extend MockedRequest and MockedResponse types

Add ExtendedMockedRequest interface with optional operationName, extensions, and context properties.
Add ExtendedMockedResponse interface with optional variableMatcher function.
Update ApolloClientParameters to use extended types for better type safety.

This resolves TypeScript errors when accessing these properties on mock objects.

hanihusam committed:
fix: update Panel components to use extended types

Update Panel.tsx and PanelContent.tsx to use ExtendedMockedResponse type instead of MockedResponse.
Remove unused MockedResponse imports and add ExtendedMockedResponse imports.

This fixes TypeScript errors when accessing operationName, extensions, and context properties.

hanihusam committed:
fix: replace ApolloError with GraphQLError in stories

Replace non-existent ApolloError import with GraphQLError from graphql package.
Update error instantiation to use GraphQLError constructor.

This fixes TypeScript import error in DisplayLocations.stories.ts.

hanihusam committed:
feat: upgrade to Apollo Client v4

- Update Apollo Client from 3.13.8 to 4.0.5
- Update peer dependency to support both v3 and v4
- Remove deprecated dependencies no longer needed in v4

hanihusam committed:
fix: update import paths for Apollo Client v4

- Change MockedProvider import to @apollo/client/testing/react
- Update useQuery import to @apollo/client/react
- Add proper TypeScript types for queries

hanihusam committed:
fix: update error handling and module declarations

- Replace ApolloError with GraphQLError in stories
- Update module declaration for storybook compatibility
- Improve TypeScript imports and formatting

hanihusam committed:
docs: update documentation and config for Apollo Client v4

- Update README with v4 import paths
- Update example error handling in documentation
- Format Storybook configuration for consistency

lifeiscontent commented on README.md:
```diff
@@ -212,7 +213,7 @@ Example.parameters = {
           query: DashboardPageQuery,
         },
-        error: new ApolloError("This is a mock network error"),
+        error: new Error("This is a mock network error"),
```
@hanihusam shouldn't this be a GraphQLError?

hanihusam replied:
You're right!. I'll replace it.

hanihusam replied:
I added more error examples based on the [doc](https://www.apollographql.com/docs/react/development-testing/testing#testing-error-states) @lifeiscontent

hanihusam committed:
docs: add comprehensive example of error testing

hanihusam committed:
chore: upgrade dependencies to latest versions

- Update @apollo/client from 4.0.5 to 4.0.7
- Update Storybook packages from 9.0.12 to 9.1.15

hanihusam committed:
refactor: simplify error handling in stories

- Replace GraphQLError with generic Error for clearer examples
- Rename WithVariableMatcher to WithDynamicVariable for clarity
- Remove unused GraphQLError import

hanihusam committed:
refactor: simplify type definitions

- Remove variableMatcher from ExtendedMockedResponse
- Simplify ApolloClientParameters type definition

hanihusam commented:
Hi @lifeiscontent any additional review for this PR?

hanihusam committed:
fix: decorators export needs an explicit type annotation
